src/analysis/seminorm.lean

```diff
@@ -1212,4 +1249,281 @@ end
 end norm
 end gauge
 
--- TODO: topology induced by family of seminorms, local convexity.
+/-! ### Topology induced by a family of seminorms -/
+
+namespace seminorm
+
+section filter_basis
+
+variables [normed_field 𝕜] [add_comm_group E] [module 𝕜 E] [decidable_eq ι] [inhabited ι]
+
+def seminorm_basis_zero (p : ι → seminorm 𝕜 E) : set (set E) :=
+  ⋃ (s : finset ι) r (hr : 0 < r), singleton $ ball (s.sup p) (0 : E) r
+
+lemma seminorm_basis_zero_iff (p : ι → seminorm 𝕜 E) (U : set E) :
+  U ∈ seminorm_basis_zero p ↔ ∃ (i : finset ι) r (hr : 0 < r), U = ball (i.sup p) 0 r :=
+by simp only [seminorm_basis_zero, mem_Union, mem_singleton_iff]
+
+lemma seminorm_basis_zero_mem (p : ι → seminorm 𝕜 E) (i : finset ι) {r : ℝ} (hr : 0 < r) :
+  (i.sup p).ball 0 r ∈ seminorm_basis_zero p :=
+(seminorm_basis_zero_iff _ _).mpr ⟨i,_,hr,rfl⟩
+
+lemma seminorm_basis_zero_singleton_mem (p : ι → seminorm 𝕜 E) (i : ι) {r : ℝ} (hr : 0 < r) :
+  (p i).ball 0 r ∈ seminorm_basis_zero p :=
+(seminorm_basis_zero_iff _ _).mpr ⟨{i},_,hr, by rw finset.sup_singleton⟩
+
+lemma seminorm_basis_zero_nonempty (p : ι → seminorm 𝕜 E) : (seminorm_basis_zero p).nonempty :=
+begin
+  refine set.nonempty_def.mpr ⟨ball (p (arbitrary ι)) 0 1, _⟩,
+  exact seminorm_basis_zero_singleton_mem _ (arbitrary ι) zero_lt_one,
+end
+
+lemma seminorm_basis_zero_intersect (p : ι → seminorm 𝕜 E)
+  (U V : set E) (hU : U ∈ seminorm_basis_zero p) (hV : V ∈ seminorm_basis_zero p) :
+  (∃ (z : set E) (H : z ∈ (seminorm_basis_zero p)), z ⊆ U ∩ V) :=
+begin
+  rcases (seminorm_basis_zero_iff p U).mp hU with ⟨ι'₁, r₁, hr₁, hU⟩,
+  rcases (seminorm_basis_zero_iff p V).mp hV with ⟨ι'₂, r₂, hr₂, hV⟩,
+  use ((ι'₁ ∪ ι'₂).sup p).ball 0 (min r₁ r₂),
+  refine ⟨seminorm_basis_zero_mem p (ι'₁ ∪ ι'₂) (lt_min_iff.mpr ⟨hr₁, hr₂⟩), _⟩,
+  rw [hU, hV, ball_finset_sup_eq_Inter _ _ _ (lt_min_iff.mpr ⟨hr₁, hr₂⟩),
+    ball_finset_sup_eq_Inter _ _ _ hr₁, ball_finset_sup_eq_Inter _ _ _ hr₂,
+    ←set.Inter_inter_distrib],
+  -- Yael probably knows how to make this look nicer
+  refine set.Inter_mono (λ i, (set.subset_inter _ _)),
+  { refine set.Inter_mono' (λ hi, _),
+    use finset.mem_of_subset (finset.subset_union_left ι'₁ ι'₂) hi,
+    exact ball_mono (min_le_of_left_le (le_refl _)) },
+  refine set.Inter_mono' (λ hi, _),
+  use finset.mem_of_subset (finset.subset_union_right ι'₁ ι'₂) hi,
+  exact ball_mono (min_le_of_right_le (le_refl _)),
```

Comment:
@YaelDillies I saw you recently refactored the `Inter_*` lemmas, is there a way to do this in a nicer way? I did not get `Inter_mono₂` to work.